File: python/wow_serialization/wow_cbor.py
"""WowCbor: CBOR wrapper with WoW-specific post-processing.

Post-processing applied after decode:
- CBOR byte strings (bytes) → UTF-8 text strings
- Sequential 1-based integer-keyed dicts → lists (Lua table convention)
"""

import cbor2


class WowCbor:
    @staticmethod
    def decode(data: bytes):
        """Decode CBOR bytes to a Python value with WoW-specific post-processing."""
        raw = cbor2.loads(data)
        return WowCbor._post_process(raw)

    @staticmethod
    def encode(data) -> bytes:
        """Encode a Python value to CBOR bytes."""
        return cbor2.dumps(data)

    @classmethod
    def _post_process(cls, val):
        if isinstance(val, bytes):
            return val.decode("utf-8", errors="replace")
        if isinstance(val, list):
            return [cls._post_process(v) for v in val]
        if isinstance(val, dict):
            result = {k: cls._post_process(v) for k, v in val.items()}
            return cls._apply_array_detection(result)
        return val

    @staticmethod
    def _apply_array_detection(d: dict):
        """Convert sequential 1-based integer-keyed dict to list."""
        if not d:
            return d
        keys = list(d.keys())
        if not all(isinstance(k, int) and k > 0 for k in keys):
            return d
        sorted_keys = sorted(keys)
        if sorted_keys != list(range(1, len(keys) + 1)):
            return d
        return [d[k] for k in sorted_keys]

File: ruby/lib/azerite/wow_cbor.rb
# typed: strict
# frozen_string_literal: true

require 'cbor'
require 'sorbet-runtime'

# WowCbor wraps the CBOR library with WoW-specific post-processing:
# - CBOR byte strings (binary Ruby strings) are decoded to UTF-8 text
# - Sequential 1-based integer-keyed maps are converted to arrays (Lua table convention)
module Azerite
  class WowCbor
    extend T::Sig

    class << self
      extend T::Sig

      # Decode CBOR bytes to a Ruby value with WoW-specific post-processing.
      sig { params(bytes: String).returns(T.untyped) }
      def decode(bytes)
        raw = CBOR.decode(bytes)
        post_process(raw)
      end

      # Encode a Ruby value to CBOR bytes.
      sig { params(data: T.untyped).returns(String) }
      def encode(data)
        data.to_cbor
      end

      private

      sig { params(val: T.untyped).returns(T.untyped) }
      def post_process(val)
        case val
        when String
          # CBOR byte strings come back as binary-encoded Ruby strings
          if [Encoding::BINARY, Encoding::ASCII_8BIT].include?(val.encoding)
            val.encode('UTF-8', invalid: :replace, undef: :replace)
          else
            val
          end
        when Array
          val.map { |v| post_process(v) }
        when Hash
          result = val.transform_values { |v| post_process(v) }
          apply_array_detection(result)
        else
          val
        end
      end

      sig { params(hash: T::Hash[T.untyped, T.untyped]).returns(T.untyped) }
      def apply_array_detection(hash)
        keys = hash.keys
        return hash if keys.empty?
        return hash unless keys.all? { |k| k.is_a?(Integer) && k.positive? }

        sorted = keys.sort
        return hash unless sorted == (1..keys.size).to_a

        sorted.map { |k| hash[k] }
      end
    end
  end
end
